working/0047. Spread Collections/spread-collections.md

```diff
@@ -0,0 +1,574 @@
+}
+command.add(mainPath);
+```
+
+To handle these gracefully, we support a `...?` "null-aware spread" operator. In
+cases where the spread expression evaluates to null, that expands to an empty
+collection instead of throwing a runtime expression.
+
+That turns the example to:
+
+```dart
+var command = [
+  engineDartPath,
+  '--target=flutter',
+  ...?extraFrontEndOptions,
+  mainPath
+];
+```
+
+More complex conditional expressions than simple null checks come up often too,
+but those are out of scope for this proposal.
+
+## Syntax
+
+We extend the list grammar to allow *spread elements* in addition to regular
+elements:
+
+```
+listLiteral:
+  const? typeArguments? '[' listElementList? ']'
+  ;
+
+listElementList:
+  listElement ( ',' listElement )* ','?
+  ;
+
+listElement:
+  expression |
+  ( '...' | '...?' ) expression
+  ;
+```
+
+Instead of `expressionList`, this uses a new `listElementList` rule since
+`expressionList` is used elsewhere in the grammar where spreads aren't allowed.
+Each element in a list is either a normal expression or a *spread element*. If
+the spread element starts with `...?`, it's a *null-aware spread element*.
+
+The changes for map literals are similar:
+
+```
+mapLiteral:
+  const? typeArguments? '{' mapLiteralEntryList? '}' ;
+
+mapLiteralEntryList:
+  mapLiteralEntry ( ',' mapLiteralEntry )* ','?
+  ;
+
+mapLiteralEntry:
+  expression ':' expression |
+  ( '...' | '...?' ) expression
+  ;
+```
+
+Note that a *spread entry* for a map is an expression, not a key/value pair.
+Similar to lists, a spread entry that starts with `...?` is a *null-aware spread
+entry*.
+
+## Static Semantics
+
+Since the spread is unpacked and its individual elements added to the containing
+collection, we don't require the spread expression *itself* to be assignable to
+the collection's type. For example, this is allowed:
+
+```dart
+var numbers = <num>[1, 2, 3];
+var ints = <int>[...numbers];
+```
+
+This works because the individual elements in `numbers` do happen to have the
+right type even though the list that contains them does not. However, the spread
+object does need to be "spreadable"&mdash;it must be some kind of Iterable for a
+list literal or a Map for a map literal.
+
+It is a static error if:
+
+*   A spread element in a list literal has a static type that is not assignable
+    to `Iterable<Object>`.
+
+*   If a list spread element's static type implements `Iterable<T>` for some `T`
+    and `T` is not assignable to the element type of the list.
+
+*   A spread element in a map literal has a static type that is not assignable
+    to `Map<Object, Object>`.
+
+*   If a map spread element's static type implements `Map<K, V>` for some `K`
+    and `V` and `K` is not assignable to the key type of the map or `V` is not
+    assignable to the value type of the map.
+
+If implicit downcasts are disabled, then the "is assignable to" parts here
+become strict subtype checks instead.
+
```

Comment:
There is a specified error preventing duplicate keys in map literals when the keys are constant.  Should we consider that for nested map literals + spread?  Should the following be an error?

```
  var a = {3 : true,
               ...{3 : false}};
```

Reply:
I thought about duplicate keys, but the static checking for that is even pretty brittle for flat literals since it only works for literal keys (obviously):

```dart
var key = "key";
var map = {key: 1, key: 2}; // <-- No error.
```

So I figured the duplicate key checking wasn't very high value and wasn't worth extending into spread. Also, it might get weird if we introduce "conditional spread" features.

Reply:
I believe the rule about duplicate keys only apply to constant maps:

> It is a compile-time error if two keys of a constant map literal are equal.

so since spreads do not apply to constant maps, that issue is moot.